plugins/slack_app/app/controllers/slack_app_controller.rb

```diff
@@ -0,0 +1,152 @@
+    # Interesting parts of params:
+    # {
+    #   "user_id"=>"U0HAGH3AB", <-- look up Samson user using this
+    #   "command"=>"/deploy",
+    #   "text"=>"foo bar baz", <-- this is what to deploy
+    #   "response_url"=>"https://hooks.slack.com/commands/T0HAGP0J2/58604540277/g0xd4K2KOsgL9zXwR4kEc0eL"
+    #    ^^^^ This is how to respond later, we can also directly render some JSON
+    # }
+    return unknown_user unless @current_user_from_slack
+
+    # Parse the command
+    projectname, branchname, stagename = @payload['text'].match(/([^\s\/]+)(?:\/(\S+))?\s*(?:to\s+(.*))?/).captures
+
+    # Sanity checks
+    begin
+      project = Project.find_by_param! projectname
+    rescue ActiveRecord::RecordNotFound
+      return unknown_project(projectname)
+    end
+
+    return unauthorized_deployer unless @current_user_from_slack.deployer_for?(project)
+
+    begin
+      stage = project.stages.find_by_param!(stagename || 'production')
+    rescue ActiveRecord::RecordNotFound
+      return unknown_stage(project, stagename)
+    end
+
+    deploy_service = DeployService.new(@current_user_from_slack)
+    deploy = deploy_service.deploy!(stage, reference: branchname || 'master')
+    DeployResponseUrl.create! deploy: deploy, response_url: @payload[:response_url]
+
+    SlackMessage.new(deploy).message_body
+  end
+
+  def interact_response
+    # This means someone clicked the "Approve" button in the channel
+    deploy = Deploy.find_by_id @payload['callback_id']
+    return '_(Whoops, this deploy disappeared.)_' unless deploy
+
+    # Is this user connected?
+    return {
+      replace_original: false,
+      text: unknown_user
+    } unless @current_user_from_slack
+
+    # Can this user approve the deployment?
+    return {
+      replace_original: false,
+      text: "You can't approve your own deploys, silly!"
+    } unless @current_user_from_slack != deploy.user
+    return {
+      replace_original: false,
+      text: "It doesn't look like you have permissions to approve that deploy."
+    } unless @current_user_from_slack.deployer_for?(deploy)
+
+    # Buddy up
+    deploy.confirm_buddy!(@current_user_from_slack)
+    SlackMessage.new(deploy).message_body
+  end
+
+  def unknown_user
+    "Sorry, I don't recognize you. Perhaps you should visit " \
+    "#{url_for action: :oauth, only_path: false} " \
+    "to connect your account."
+  end
+
+  def unknown_project(project)
+    "Sorry, I don't know a project called `#{project}`. Try your project's permalink."
+  end
+
+  def unknown_stage(project, stagename)
+    "Sorry, " \
+    "<#{url_for controller: :projects, action: :show, id: project, only_path: false}|#{project.name}> " \
+    "doesn't have a stage named `#{stagename}`."
+  end
+
+  def unauthorized_deployer(project)
+    "Sorry, it doesn't look like you have permission to create a deployment" \
+    "on `#{project}`."
+  end
+
+  def slack
+    @slack_token ||= SlackIdentifier.app_token
+    Slack.configure do |config|
```

Comment:
we cannot change the global token on every request ... so maybe do something like `with_slack_token do` that resets it after